processor/src/processingManager.js

```diff
@@ -0,0 +1,233 @@
+
+    return asyncBackoff(onBackoff, onFail, MAXIMUM_NUM_OF_RETRIES)
+  }
+
+  const { id, isEmpty, isMirror } = await queryGiteaRepoWithBackoff(repoQuery)
+
+  return { id, isMirror, isEmpty }
+}
+
+/**
+ *
+ * @param {string} repoId
+ * @returns {Promise<number>}
+ */
+async function queryGiteaRepoMigrationStatus(repoId) {
+  // See gitea/models/task.go
+  // https://github.com/kitspace/gitea/blob/d1342b36f1b5e4ac0271d0f01c5270cd087f3575/models/task.go#L22
+  const MigrationTaskType = 0
+  const migrationStatusQuery = {
+    name: 'fetch-migration-status',
+    text: 'select status from task where repo_id=$1 and type=$2',
+    values: [repoId, MigrationTaskType],
+  }
+
+  /**
+   * Query the migration task status for a gitea repo, with exponential backoff.
+   * @returns
+   */
+  const queryMigrationStatusWithBackoff = async migrationStatusQuery => {
+    const onBackoff = async (num, delay, resolve) => {
+      log.debug(
+        'Backoff started, querying migration status for repo' +
+          `Id(${migrationStatusQuery.values[0]}): ${num} ${delay}ms`,
+      )
+
+      const migrationStatusQueryResult = await client.query(migrationStatusQuery)
+      if (migrationStatusQueryResult.rows.length === 1) {
+        const { status } = migrationStatusQueryResult.rows[0]
+        log.debug(`Repo: Id(${repoId})'s migration status: ${status}.`)
+        return resolve(status)
+      }
+    }
+
+    const onFail = reject =>
+      reject(
+        new Error(
+          `Repo: ${repoId} migration task not found after ${MaximumRetries} trials.`,
+        ),
+      )
+
+    return asyncBackoff(onBackoff, onFail, MAXIMUM_NUM_OF_RETRIES)
+  }
+
+  const repoMigrationStatus = await queryMigrationStatusWithBackoff(
+    migrationStatusQuery,
+  )
+
+  return repoMigrationStatus
+}
+
+/**
+ *
+ * @typedef {Object} ExponentialOptions
+ * @property {number=} randomisationFactor
+ * @property {number=} initialDelay
+ * @property {number=} maxDelay
+ * @property {number=} factor
+ *
+ * @param {function(number,  number, function): Promise<void>} onBackoff
+ * @param {function(): Promise<void>} onFail
+ * @param {number} maximumRetries
+ * @param {ExponentialOptions=} config
+ * @returns
+ */
+async function asyncBackoff(onBackoff, onFail, maximumRetries, config = {}) {
+  const backoffInstance = backOff({
+    ...config,
+    initialDelay: BACKOFF_INITIAL_DELAY,
+  })
+  backoffInstance.failAfter(maximumRetries)
+
+  return new Promise((resolve, reject) => {
+    const resetBackoffAndResolve = value => {
+      backoffInstance.reset()
+      resolve(value)
+    }
+
+    backoffInstance
+      .on(
+        'backoff',
+        async (num, delay) => await onBackoff(num, delay, resetBackoffAndResolve),
+      )
+      .on('ready', () => backoffInstance.backoff())
+      .on('fail', async () => await onFail(reject))
+      .backoff()
+  })
+}
+
+const ProcessingManager = {
+  client,
+  isRepoReadyForProcessing,
+}
+
+module.exports = { ProcessingManager }
```

Comment:
We need a better name for all of this. Name it after what it's doing, it's definitely not "managing" the processing.

Reply:
Renamed it to `isReadyForPorcessing`.